sdk/src/account.rs

```diff
@@ -538,17 +538,39 @@ impl Account {
 }
 
 impl AccountSharedData {
-    pub fn set_data_from_slice(&mut self, data: &[u8]) {
-        let len = self.data.len();
-        let len_different = len != data.len();
-        let different = len_different || data != &self.data[..];
-        if different {
-            self.data = Arc::new(data.to_vec());
-        }
+    pub fn set_data_from_slice(&mut self, new_data: &[u8]) {
+        let data = match Arc::get_mut(&mut self.data) {
+            // The buffer isn't shared, so we're going to memcpy in place.
+            Some(data) => data,
+            // If the buffer is shared, the cheapest thing to do is to clone the
+            // incoming slice and replace the buffer.
+            None => return self.set_data(new_data.to_vec()),
+        };
+
+        let new_len = new_data.len();
+
+        // Reserve additional capacity if needed. Here we make the assumption
+        // that growing/shrinking the current buffer is cheaper than doing a
+        // whole new allocation to make `new_data` owned.
+        //
+        // This assumption holds true during CPI, especially when the account
+        // size doesn't change but the account is only changed in place. And
+        // it's also true when the account is grown/srunk by a small margin (the
+        // realloc limit is quite low), in which case the allocator can just
+        // update the allocation metadata without moving.
+        data.reserve(new_len.saturating_sub(data.len()));
+
+        // Safety:
+        // we just explicitly reserved at least new_len capacity and we're going
+        // to write into the whole range so it's all going to be initialized
+        unsafe { data.set_len(new_len) };
+        data[..new_len].copy_from_slice(new_data);
```

Comment:
Why this half specified bound `..new_len` and not `0..new_len` no indexing at all?

Reply:
correct indexing is not needed, copy pasta from a previous version where I was truncating

Reply:
Ok, then i would clean it up. Also, I think you need a `#[allow(clippy::uninit_vec)` to get through the tests.